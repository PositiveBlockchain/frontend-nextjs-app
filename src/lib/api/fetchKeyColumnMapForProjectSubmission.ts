import { fetchSheetData } from "../google/sheets";
import { config } from "@/configuration/config";

const {
  SPREADSHEET_ID,
  sheets: { pendingProjects },
} = config.constants.google.sheets.databaseSheet;

export async function fetchKeyColumnMapForProjectSubmission(): Promise<
  Record<string, number>
> {
  const { keyRow } = pendingProjects;
  const range = `${pendingProjects.name}!A${keyRow}:${keyRow}`;

  // Fetch the sheet data for data from keyRow to headerRow
  const sheetData = await fetchSheetData({
    spreadsheetId: SPREADSHEET_ID,
    range,
  });

  if (!sheetData) throw new Error("No sheet data found");

  return sheetDataToKeyMap(sheetData);
}

/**
 * Iterate through the sheet data and create a map of key to column index
 * @param sheetData
 * @returns
 */
function sheetDataToKeyMap(sheetData: string[][]) {
  const keyMap: { [key: string]: number } = {};
  const [...keys] = sheetData[0];

  keys.forEach((key, index) => {
    // Ignore empty keys
    if (key) {
      keyMap[key] = index;
    }
  });

  return keyMap;
}
