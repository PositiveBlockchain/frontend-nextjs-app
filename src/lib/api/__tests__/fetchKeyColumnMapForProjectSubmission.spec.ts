import { fetchKeyColumnMapForProjectSubmission } from "../fetchKeyColumnMapForProjectSubmission";

describe("fetchKeyColumnMapForProjectSubmission", () => {
  let types: any = [];

  beforeAll(async () => {
    types = await fetchKeyColumnMapForProjectSubmission();
  });

  it("should match snapshot", () => {
    expect(types).toMatchSnapshot();
  });

  it("should return the column index map", async () => {
    expect(types).toMatchObject({
      PUBLIC_project_name: 3,
      PUBLIC_active: 4,
    });
  });
});
