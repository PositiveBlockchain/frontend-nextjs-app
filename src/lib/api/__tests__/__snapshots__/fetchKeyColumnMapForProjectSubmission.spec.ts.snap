// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchKeyColumnMapForProjectSubmission should match snapshot 1`] = `
{
  "#REF!": 52,
  "PUBLIC_active": 4,
  "PUBLIC_blockchain_technology": 33,
  "PUBLIC_blockchain_type": 32,
  "PUBLIC_business_tagline": 17,
  "PUBLIC_categories_list": 14,
  "PUBLIC_description_short_value_proposition_in_a_tweet": 15,
  "PUBLIC_discord_url": 48,
  "PUBLIC_facebook_url": 46,
  "PUBLIC_founder_names_list": 20,
  "PUBLIC_hq_region": 25,
  "PUBLIC_hq_subregion": 26,
  "PUBLIC_linkedin_url": 47,
  "PUBLIC_logo_url": 35,
  "PUBLIC_long_description": 16,
  "PUBLIC_main_category": 6,
  "PUBLIC_organization_type": 27,
  "PUBLIC_other_links_list": 53,
  "PUBLIC_pb_partner_tag": 36,
  "PUBLIC_primary_headquarter_city": 21,
  "PUBLIC_primary_headquarter_country": 22,
  "PUBLIC_project_name": 3,
  "PUBLIC_secondary_headquarter_city": 23,
  "PUBLIC_secondary_headquarter_country": 24,
  "PUBLIC_servicing_area": 29,
  "PUBLIC_servicing_region": 30,
  "PUBLIC_sponsors_partners_list": 28,
  "PUBLIC_sub_categories_list": 13,
  "PUBLIC_subregions_list": 31,
  "PUBLIC_tags_keywords_list": 18,
  "PUBLIC_token_ticker": 34,
  "PUBLIC_twitter_url": 45,
  "PUBLIC_video_url": 49,
  "PUBLIC_website": 5,
  "PUBLIC_white_paper_url": 50,
  "PUBLIC_year_creation": 19,
  "This key needs to correspond to a field in the database if it is to be easily merged": 1,
  "additional_comments": 55,
  "agrees_to_data_privacy_agreement": 43,
  "category_1": 7,
  "category_2": 8,
  "category_3": 9,
  "is_positive_blockchain_project": 2,
  "is_project_owner": 37,
  "key": 0,
  "should_receive_newsletter": 42,
  "slug": 54,
  "sub_category_1": 10,
  "sub_category_2": 11,
  "sub_category_3": 12,
  "submitted_at": 38,
  "submitted_at_ISO": 39,
  "submitted_by_email": 41,
  "submitter_name": 40,
}
`;
